DCCpp_Uno/src/PacketRegister.cpp

```diff
@@ -277,37 +278,39 @@ void RegisterList::writeCVByte(const char *s) volatile{
 
   if(sscanf(s,"%d %d %d %d",&cv,&bValue,&callBack,&callBackSub)!=4)          // cv = 1-1024
     return;
-  cv--;                              // actual CV addresses are cv-1 (0-1023)
+  cv--;                                  // actual CV addresses are cv-1 (0-1023)
 
   bWrite[0]=0x7C+(highByte(cv)&0x03);   // any CV>1023 will become modulus(1024) due to bit-mask of 0x03
   bWrite[1]=lowByte(cv);
   bWrite[2]=bValue;
 
-  loadPacket(0,resetPacket,2,1);
-  loadPacket(0,bWrite,3,4);
-  loadPacket(0,resetPacket,2,1);
-  loadPacket(0,idlePacket,2,10);
-
+ 
+  loadPacket(0,resetPacket,2,3);        // NMRA recommends starting with 3 reset packets
+  loadPacket(0,bWrite,3,5);             // NMRA recommends 5 verify packets
+  loadPacket(0,bWrite,3,6);             // NMRA recommends 6 write or reset packets for decoder recovery time
```

Comment:
should we send an idle packet instead of another write packet here as done in the read sections?

Reply:
I'm wondering the same. If you look at the code from trainboard that "fixed" many of the issues, WriteCVByte went from this:
```
loadPacket(0,resetPacket,2,1);
loadPacket(0,bWrite,3,4);
loadPacket(0,resetPacket,2,1);
loadPacket(0,idlePacket,2,10);

```
to this:

```
loadPacket(0,resetPacket,2,1);   
loadPacket(0,bWrite,3,4);        
loadPacket(0,idlePacket,2,10);
```

But the next code that does the verify is different and they changed the last line from this:

`loadPacket(0,resetPacket,2,1); `

to 

`loadPacket(0,bWrite,3,1)`

So to synopsize; If I have it strait, from writeCVByte: They went from "**reset, write, reset, idle**" to "**reset, write, idle**" and we are making it "**reset, write, write**"

and on the verify, They went to from "**reset, write, reset**" to "**reset, write, write**" which is what I believe we have now. And of course we also changed parameters like increasing the number of packets. Wish I had a scope to see what it generates. What do you think?

Reply:
The problem with the original code is the reset is processed before the ACKs are checked. The reset needs to be sent after the ACK check.

I'd suggest the flow for writeCVByte be: reset, write, idle, CHECK ACK, reset.

As for using a scope, that would only show you the packets as sent and not the behavior of the decoder which processes the packets. The decoder is supposed to stop what it is doing and reset it's state when it receives a reset packet. It it has not sent an ACK *AND* the base station has not checked for an ACK then it is potentially going to be missed.

Reply:
NMRA specs allow a write or reset for recovery time.  
Optional Power-On-Cycle, if needed
[ 3 or more Reset Packets
[ 5 or more Verify packets to a single CV 1-1023,
followed by 1 or more Reset Packets, if an
acknowledgement is detected ]
 Or 5 or more Writes to a single CV 1-1023
] 6 or more Identical Write or Reset packets (DecoderRecovery-Time) ]


So I think the Reset, write, write, Ack, reset I think is whats in there now and the reset is after the ack loop No?

Reply:
That should be fine for now. One other aspect that is out of spec is the fact that the PROG track is always energized rather than energized on demand, that should likely be tracked separately.